File: game-engine/src/test/java/org/luxons/sevenwonders/game/effects/ScienceProgressTest.java
package org.luxons.sevenwonders.game.effects;

import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.boards.Science;
import org.luxons.sevenwonders.game.boards.ScienceType;
import org.luxons.sevenwonders.game.resources.ResourceType;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;

@RunWith(Theories.class)
public class ScienceProgressTest {

    @DataPoints
    public static int[] elementsCount() {
        return new int[] {0, 1, 2};
    }

    @Theory
    public void apply_initContainsAddedScience(int initCompasses, int initWheels, int initTablets, int initJokers,
            int compasses, int wheels, int tablets, int jokers) {
        Board board = TestUtilsKt.testBoard(ResourceType.ORE);
        Science initialScience = TestUtilsKt.createScience(initCompasses, initWheels, initTablets, initJokers);
        board.getScience().addAll(initialScience);

        ScienceProgress effect = TestUtilsKt.createScienceProgress(compasses, wheels, tablets, jokers);
        effect.apply(board);

        assertEquals(initCompasses + compasses, board.getScience().getQuantity(ScienceType.COMPASS));
        assertEquals(initWheels + wheels, board.getScience().getQuantity(ScienceType.WHEEL));
        assertEquals(initTablets + tablets, board.getScience().getQuantity(ScienceType.TABLET));
        assertEquals(initJokers + jokers, board.getScience().getJokers());
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/effects/ScienceProgressTest.kt
package org.luxons.sevenwonders.game.effects

import org.junit.Assert.assertEquals
import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.runner.RunWith
import org.luxons.sevenwonders.game.boards.ScienceType
import org.luxons.sevenwonders.game.resources.ResourceType
import org.luxons.sevenwonders.game.test.createScience
import org.luxons.sevenwonders.game.test.createScienceProgress
import org.luxons.sevenwonders.game.test.testBoard

@RunWith(Theories::class)
class ScienceProgressTest {

    @Theory
    fun apply_initContainsAddedScience(
        initCompasses: Int,
        initWheels: Int,
        initTablets: Int,
        initJokers: Int,
        compasses: Int,
        wheels: Int,
        tablets: Int,
        jokers: Int
    ) {
        val board = testBoard(ResourceType.ORE)
        val initialScience = createScience(initCompasses, initWheels, initTablets, initJokers)
        board.science.addAll(initialScience)

        val effect = createScienceProgress(compasses, wheels, tablets, jokers)
        effect.applyTo(board)

        assertEquals(initCompasses + compasses, board.science.getQuantity(ScienceType.COMPASS))
        assertEquals(initWheels + wheels, board.science.getQuantity(ScienceType.WHEEL))
        assertEquals(initTablets + tablets, board.science.getQuantity(ScienceType.TABLET))
        assertEquals(initJokers + jokers, board.science.jokers)
    }

    companion object {

        @JvmStatic
        @DataPoints
        fun elementsCount(): IntArray {
            return intArrayOf(0, 1, 2)
        }
    }
}
